File: apps/mobile/ios/Sources/FlagshipUI/Screens/TierStatusScreen.swift
import SwiftUI
import FlagshipAPI
import FlagshipCore

/// P7 — dedicated tier-status / subscription screen. Mirrors the
/// canonical webapp `views/tier-status.js`: a tier badge, LLM credits
/// (today + lifetime), dispatcher usage with a progress bar (usage vs
/// free quota), the custom-domains list, and the reserved-names list.
public struct TierStatusScreen: View {
    @Environment(\.colorScheme) private var scheme
    @State private var vm: TierStatusViewModel

    public init(vm: TierStatusViewModel) {
        _vm = State(initialValue: vm)
    }

    public var body: some View {
        let c = FSColors.scheme(scheme)
        ScrollView {
            VStack(alignment: .leading, spacing: FS.space.s6) {
                switch vm.state {
                case .idle, .loading:
                    ForEach(0..<2, id: \.self) { _ in ServerCardSkeleton() }
                case .failed(let msg):
                    ErrorCard(message: msg)
                case .loaded(let t):
                    tierCard(t, c: c)
                    llmSection(t, c: c)
                    dispatcherSection(t, c: c)
                    customDomainsSection(t, c: c)
                    reservedNamesSection(t, c: c)
                }
                Spacer().frame(height: FS.space.s12)
            }
            .padding(.horizontal, FS.space.s6)
            .padding(.top, FS.space.s4)
        }
        .background(c.bg.ignoresSafeArea())
        .navigationTitle("Tier & usage")
        .navigationBarTitleDisplayMode(.inline)
        .refreshable { await vm.load() }
        .task {
            if case .idle = vm.state { await vm.load() }
        }
    }

    // MARK: - Sections (1:1 with tier-status.js)

    private func tierCard(_ t: TierStatusResponse, c: FSColors) -> some View {
        FSCard {
            HStack {
                Text("Tier").foregroundColor(c.textMuted)
                Spacer()
                FSPill(tierLabel(t.tier), kind: t.tier == "free" ? .idle : .online)
                    .accessibilityIdentifier("tier-status-badge")
            }
        }
    }

    @ViewBuilder
    private func llmSection(_ t: TierStatusResponse, c: FSColors) -> some View {
        sectionHeader("LLM credits", c: c)
        FSCard {
            VStack(spacing: FS.space.s2) {
                if let day = t.llmCreditsRemainingDay {
                    row("today remaining", grouped(day), c: c)
                } else {
                    row("today", "— (BYOK or promo not in use)", c: c)
                }
                if let total = t.llmCreditsRemainingTotal {
                    row("lifetime remaining", grouped(total), c: c)
                }
            }
        }
    }

    @ViewBuilder
    private func dispatcherSection(_ t: TierStatusResponse, c: FSColors) -> some View {
        sectionHeader("Dispatcher relay", c: c)
        FSCard {
            if let used = t.dispatcherUsageGBmonth {
                VStack(alignment: .leading, spacing: FS.space.s2) {
                    HStack {
                        Text("this month").foregroundColor(c.textMuted)
                        Spacer()
                        Text(dispatcherValue(used: used, quota: t.dispatcherFreeQuotaGBmonth))
                            .foregroundColor(c.text)
                    }
                    progressBar(
                        percent: TierStatusViewModel.usagePercent(
                            used: used, quota: t.dispatcherFreeQuotaGBmonth
                        ),
                        c: c
                    )
                }
            } else {
                row("usage", "—", c: c)
            }
        }
    }

    @ViewBuilder
    private func customDomainsSection(_ t: TierStatusResponse, c: FSColors) -> some View {
        sectionHeader("Custom domains", c: c)
        if t.customDomains.isEmpty {
            placeholderCard("none — your default subdomain is forever-free", c: c)
        } else {
            ForEach(t.customDomains, id: \.self) { d in
                FSCard { Text(d).font(FS.font.mono()).foregroundColor(c.text) }
            }
        }
    }

    @ViewBuilder
    private func reservedNamesSection(_ t: TierStatusResponse, c: FSColors) -> some View {
        sectionHeader("Reserved names", c: c)
        if t.reservedNames.isEmpty {
            placeholderCard("none — your username is FCFS-free", c: c)
        } else {
            ForEach(t.reservedNames, id: \.self) { n in
                FSCard { Text(n).font(FS.font.mono()).foregroundColor(c.text) }
            }
        }
    }

    // MARK: - Bits

    private func tierLabel(_ tier: String) -> String {
        switch tier {
        case "free":  return "free"
        case "promo": return "promo"
        case "byok":  return "BYOK"
        default:      return tier
        }
    }

    private func dispatcherValue(used: Double, quota: Double?) -> String {
        let usedStr = String(format: "%.2f GB", used)
        if let quota {
            return "\(usedStr) / \(String(format: "%.0f", quota)) GB free"
        }
        return "\(usedStr) / —"
    }

    private func grouped(_ n: Int64) -> String {
        let f = NumberFormatter()
        f.numberStyle = .decimal
        return f.string(from: NSNumber(value: n)) ?? "\(n)"
    }

    private func progressBar(percent: Int, c: FSColors) -> some View {
        GeometryReader { geo in
            ZStack(alignment: .leading) {
                Capsule().fill(c.surfaceSunken)
                Capsule().fill(c.primary)
                    .frame(width: geo.size.width * CGFloat(percent) / 100)
            }
        }
        .frame(height: 8)
        .accessibilityIdentifier("tier-status-dispatcher-progress")
    }

    private func sectionHeader(_ title: String, c: FSColors) -> some View {
        Text(title).font(FS.font.h3()).foregroundColor(c.text)
    }

    private func placeholderCard(_ text: String, c: FSColors) -> some View {
        FSCard { Text(text).font(FS.font.bodySm()).foregroundColor(c.textMuted) }
    }

    private func row(_ label: String, _ value: String, c: FSColors) -> some View {
        HStack {
            Text(label).foregroundColor(c.textMuted)
            Spacer()
            Text(value).foregroundColor(c.text)
        }
    }
}

File: apps/mobile/android/app/src/main/java/com/flagshipserver/app/ui/screens/TierStatusScreen.kt
// P7 — dedicated tier-status / subscription screen.
//
// Mirrors the canonical webapp `views/tier-status.js` + iOS
// FlagshipUI/Screens/TierStatusScreen.swift 1:1:
//
//   - tier badge (free / promo / BYOK pill)
//   - LLM credits (today + lifetime remaining)
//   - Dispatcher relay usage card with progress bar (usage vs free quota)
//   - Custom-domains list (else "none — your default subdomain is
//     forever-free")
//   - Reserved-names list (else "none — your username is FCFS-free")

package com.flagshipserver.app.ui.screens

import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.PaddingValues
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.verticalScroll
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.collectAsState
import androidx.compose.runtime.getValue
import androidx.compose.runtime.remember
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.semantics.testTag
import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.text.font.FontFamily
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import androidx.navigation.NavController
import com.flagshipserver.app.api.TierStatusResponse
import com.flagshipserver.app.core.LocalScreensClient
import com.flagshipserver.app.ui.components.FSCard
import com.flagshipserver.app.ui.components.FSPill
import com.flagshipserver.app.ui.components.FSPillKind
import com.flagshipserver.app.ui.theme.FS
import com.flagshipserver.app.viewmodels.LoadingState
import com.flagshipserver.app.viewmodels.TierStatusViewModel
import java.text.NumberFormat
import java.util.Locale

@Composable
fun TierStatusScreen(@Suppress("UNUSED_PARAMETER") nav: NavController) {
    val client = LocalScreensClient.current
    val vm = remember { TierStatusViewModel(client) }
    val state by vm.state.collectAsState()

    LaunchedEffect(Unit) { vm.load() }

    val scroll = rememberScrollState()
    Column(
        Modifier
            .fillMaxSize()
            .verticalScroll(scroll)
            .padding(horizontal = FS.space.s6),
    ) {
        Spacer(Modifier.height(FS.space.s8))
        Text(
            "Tier & usage",
            color = FS.colors.text,
            style = TextStyle(fontSize = 28.sp, lineHeight = 36.sp, fontWeight = FontWeight.Medium),
        )
        Text(
            "Your subscription, LLM credits, dispatcher usage, custom domains, and reserved names.",
            color = FS.colors.textMuted,
            style = TextStyle(fontSize = 14.sp),
        )
        Spacer(Modifier.height(FS.space.s4))

        when (val s = state) {
            is LoadingState.Loaded -> Body(s.value)
            is LoadingState.Failed -> ErrorCard(s.message, onRetry = { vm.load() })
            else -> {
                ServerCardSkeleton()
                Spacer(Modifier.height(FS.space.s2))
                ServerCardSkeleton()
            }
        }
        Spacer(Modifier.height(FS.space.s12))
    }
}

@Composable
private fun Body(t: TierStatusResponse) {
    TierBadgeCard(t)
    Spacer(Modifier.height(FS.space.s4))

    SectionHeader("LLM credits")
    Spacer(Modifier.height(FS.space.s2))
    LlmCard(t)

    Spacer(Modifier.height(FS.space.s4))
    SectionHeader("Dispatcher relay")
    Spacer(Modifier.height(FS.space.s2))
    DispatcherCard(t)

    Spacer(Modifier.height(FS.space.s4))
    SectionHeader("Custom domains")
    Spacer(Modifier.height(FS.space.s2))
    if (t.customDomains.isEmpty()) {
        PlaceholderCard("none — your default subdomain is forever-free")
    } else {
        t.customDomains.forEach { d ->
            FSCard(padding = PaddingValues(FS.space.s4)) {
                Text(d, color = FS.colors.text, style = TextStyle(fontSize = 14.sp, fontFamily = FontFamily.Monospace))
            }
            Spacer(Modifier.height(FS.space.s2))
        }
    }

    Spacer(Modifier.height(FS.space.s4))
    SectionHeader("Reserved names")
    Spacer(Modifier.height(FS.space.s2))
    if (t.reservedNames.isEmpty()) {
        PlaceholderCard("none — your username is FCFS-free")
    } else {
        t.reservedNames.forEach { n ->
            FSCard(padding = PaddingValues(FS.space.s4)) {
                Text(n, color = FS.colors.text, style = TextStyle(fontSize = 14.sp, fontFamily = FontFamily.Monospace))
            }
            Spacer(Modifier.height(FS.space.s2))
        }
    }
}

@Composable
private fun TierBadgeCard(t: TierStatusResponse) {
    FSCard(padding = PaddingValues(FS.space.s4)) {
        Row(
            modifier = Modifier
                .fillMaxWidth()
                .semantics { testTag = "tier-status-badge-row" },
            verticalAlignment = Alignment.CenterVertically,
        ) {
            Text("Tier", color = FS.colors.textMuted, style = TextStyle(fontSize = 14.sp))
            Spacer(Modifier.weight(1f))
            FSPill(
                label = tierLabel(t.tier),
                kind = if (t.tier == "free") FSPillKind.Idle else FSPillKind.Online,
            )
        }
    }
}

@Composable
private fun LlmCard(t: TierStatusResponse) {
    FSCard(padding = PaddingValues(FS.space.s4)) {
        Column(verticalArrangement = Arrangement.spacedBy(FS.space.s2)) {
            val day = t.llmCreditsRemainingDay
            if (day != null) {
                LabelValueRow("today remaining", grouped(day))
            } else {
                LabelValueRow("today", "— (BYOK or promo not in use)")
            }
            t.llmCreditsRemainingTotal?.let { total ->
                LabelValueRow("lifetime remaining", grouped(total))
            }
        }
    }
}

@Composable
private fun DispatcherCard(t: TierStatusResponse) {
    FSCard(padding = PaddingValues(FS.space.s4)) {
        val used = t.dispatcherUsageGBmonth
        if (used != null) {
            Column(verticalArrangement = Arrangement.spacedBy(FS.space.s2)) {
                LabelValueRow("this month", dispatcherValue(used, t.dispatcherFreeQuotaGBmonth))
                ProgressBar(percent = TierStatusViewModel.usagePercent(used, t.dispatcherFreeQuotaGBmonth))
            }
        } else {
            LabelValueRow("usage", "—")
        }
    }
}

@Composable
private fun ProgressBar(percent: Int) {
    val fraction = percent.coerceIn(0, 100) / 100f
    Box(
        Modifier
            .fillMaxWidth()
            .height(8.dp)
            .clip(RoundedCornerShape(50))
            .background(FS.colors.surfaceSunken)
            .semantics { testTag = "tier-status-dispatcher-progress" },
    ) {
        if (fraction > 0f) {
            Box(
                Modifier
                    .fillMaxWidth(fraction)
                    .height(8.dp)
                    .clip(RoundedCornerShape(50))
                    .background(FS.colors.primary),
            )
        }
    }
}

@Composable
private fun SectionHeader(title: String) {
    Text(title, color = FS.colors.text, style = TextStyle(fontSize = 18.sp, fontWeight = FontWeight.SemiBold))
}

@Composable
private fun PlaceholderCard(text: String) {
    FSCard(padding = PaddingValues(FS.space.s4)) {
        Text(text, color = FS.colors.textMuted, style = TextStyle(fontSize = 13.sp))
    }
}

@Composable
private fun LabelValueRow(label: String, value: String) {
    Row(
        modifier = Modifier.fillMaxWidth(),
        verticalAlignment = Alignment.CenterVertically,
    ) {
        Text(label, color = FS.colors.textMuted, style = TextStyle(fontSize = 14.sp))
        Spacer(Modifier.weight(1f))
        Text(value, color = FS.colors.text, style = TextStyle(fontSize = 14.sp))
    }
}

private fun tierLabel(tier: String): String = when (tier) {
    "free"  -> "free"
    "promo" -> "promo"
    "byok"  -> "BYOK"
    else    -> tier
}

private fun dispatcherValue(used: Double, quota: Double?): String {
    val usedStr = String.format(Locale.US, "%.2f GB", used)
    return if (quota != null) {
        "$usedStr / ${String.format(Locale.US, "%.0f", quota)} GB free"
    } else {
        "$usedStr / —"
    }
}

private fun grouped(n: Long): String =
    NumberFormat.getIntegerInstance(Locale.US).format(n)
